internal/wasmldr/src/workload.rs

```diff
@@ -0,0 +1,224 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use log::{debug, warn};
+use wasmtime_wasi::sync::WasiCtxBuilder;
+
+/// The error codes of workload execution.
+// clippy doesn't like how "ConfigurationError" ends with "Error", so..
+#[allow(clippy::enum_variant_names)]
+// TODO: use clippy-approved names when we rework these and refactor run();
+//       until then
+#[derive(Debug)]
+pub enum Error {
+    /// configuration error
+    ConfigurationError,
+    /// export not found
+    ExportNotFound,
+    /// module instantiation failed
+    InstantiationFailed,
+    /// call failed
+    CallFailed,
+    /// I/O error
+    IoError(std::io::Error),
+    /// WASI error
+    WASIError(wasmtime_wasi::Error),
+    /// Arguments or environment too large
+    StringTableError,
+}
+
+impl From<std::io::Error> for Error {
+    fn from(err: std::io::Error) -> Self {
+        Self::IoError(err)
+    }
+}
+
+impl From<wasmtime_wasi::Error> for Error {
+    fn from(err: wasmtime_wasi::Error) -> Self {
+        Self::WASIError(err)
+    }
+}
+
+/// Result type used throughout the library.
+pub type Result<T> = std::result::Result<T, Error>;
+
+/// Runs a WebAssembly workload.
+// TODO: refactor this into multiple steps
+// Since we're not bundling the launch/deployment config into `bytes`, the
+// naive solution would just be to add new arguments for those things, like
+// WasmFeatures, stdio handling, etc - but that gets messy quick.
+// Instead we should probably refactor this into distinct steps, each with
+// its own config options (and error variants - see above).
+pub fn run<T: AsRef<str>, U: AsRef<str>>(
+    bytes: impl AsRef<[u8]>,
+    args: impl IntoIterator<Item = T>,
+    envs: impl IntoIterator<Item = (U, U)>,
+) -> Result<Box<[wasmtime::Val]>> {
+    debug!("configuring wasmtime engine");
+    let mut config = wasmtime::Config::new();
+    // Support module-linking (https://github.com/webassembly/module-linking)
+    config.wasm_module_linking(true);
+    // module-linking requires multi-memory
+    config.wasm_multi_memory(true);
+
+    // Prefer dynamic memory allocation style over static memory
+    config.static_memory_maximum_size(0);
+    config.static_memory_guard_size(0);
+    config.dynamic_memory_guard_size(0);
+    config.dynamic_memory_reserved_for_growth(1 * 1024 * 1024);
```

Comment:
We could use [`nbytes::bytes!`](https://github.com/enarx/nbytes) here.
```suggestion
    config.dynamic_memory_reserved_for_growth(bytes![1; MiB]);
```